contracts/src/v0.8/ccip/applications/external/CCIPClient.sol

```diff
@@ -0,0 +1,127 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import {IRouterClient} from "../../interfaces/IRouterClient.sol";
+
+import {Client} from "../../libraries/Client.sol";
+import {CCIPReceiver} from "./CCIPReceiver.sol";
+
+import {IERC20} from "../../../vendor/openzeppelin-solidity/v4.8.3/contracts/token/ERC20/IERC20.sol";
+import {SafeERC20} from "../../../vendor/openzeppelin-solidity/v4.8.3/contracts/token/ERC20/utils/SafeERC20.sol";
+
+/// @title CCIPClient
+/// @notice This contract implements logic for sending and receiving CCIP Messages. It utilizes CCIPReceiver's defensive patterns by default.
+/// @dev CCIPReceiver and CCIPSender cannot be simultaneously imported due to similar parents so CCIPSender functionality has been duplicated.
+contract CCIPClient is CCIPReceiver {
+  using SafeERC20 for IERC20;
+
+  event MessageSent(bytes32 messageId);
+  event FeeTokenUpdated(address oldFeeToken, address newFeeToken);
+
+  IERC20 internal s_feeToken;
+
+  /// @notice Pre-funding indicates that the contract should use funds already in its posession to pay CCIP fees.
+  /// If prefunding is disabled then fee tokens will need to be transferred from the msg.sender before sending a message.
+  bool internal immutable i_isPreFunded;
+
+  /// @dev A check for the zero-address is not explicitly performed since it is included in the CCIPBase parent constructor.
+  /// @param feeToken An ERC20-compatible token, address(0) if fees should be paid in native tokens instead.
+  constructor(address router, IERC20 feeToken, bool usePreFunding) CCIPReceiver(router) {
+    s_feeToken = feeToken;
+    i_isPreFunded = usePreFunding;
+
+    // Approve the router to spend the fee token.
+    if (address(feeToken) != address(0)) {
+      IERC20(feeToken).safeApprove(s_ccipRouter, type(uint256).max);
+    }
+  }
+
+  function typeAndVersion() external pure virtual override returns (string memory) {
```

Comment:
This probably doesn't make a lot of sense, as we would always want them to override it. Maybe good to not force typeAndVersion on them at all?